pyrolite/mineral/normative.py

```diff
@@ -137,95 +133,233 @@ def endmember_decompose(
+    mol_weights = pd.Series([60.08,79.90,101.96,159.6882,71.844,70.94,40.30,56.08,61.98,94.20,141.94, 44.01, 80.06], index = columns)
+
+    # If want to convert to FeOtotal
+#     data['FeO'] = data['FeO'] + (data['Fe2O3']*(2*71.8444/159.6882))
+#     data['Fe2O3'] = 0
+
+    # Normalise to 100 on anhydrous basis
+    totals = data.sum(axis = 1)
+    data = data.divide(totals/100, axis = 0)
+
+    # Convert to moles
+    data = data.div(mol_weights)
+
+    # Calculate corrected molecular weights for later
+    # When updating for trace elements will need to add more here
+    xFeO = data['FeO']/(data['FeO']+data['MnO'])
+    xMnO = data['MnO']/(data['FeO']+data['MnO'])
+    MWFeOcorr = (xMnO*mol_weights['MnO'])+(xFeO*mol_weights['FeO'])
+#     MWFeOcorr = (mol_weights['MnO']*(data['MnO']/(data['MnO']+data['FeO']))+(mol_weights['FeO']*data['FeO']/(data['MnO']+data['FeO'])))
+    MWil = MWFeOcorr + 79.8658
+    MWhyp = MWFeOcorr + 60.0843
+    MWhyp = (100.39*(data['MgO']/(data['FeO']+data['MgO'])))+(MWhyp*(data['FeO']/(data['FeO']+data['MgO'])))
+    MWol = (2*MWFeOcorr) + 60.0843
+    MWol = (140.69*(data['MgO']/(data['FeO']+data['MgO'])))+(MWol*(data['FeO']/(data['FeO']+data['MgO'])))
+    # MWhyp = (55.845+24.305)+(((28.085)+(16*3))*2)
+    MWmt = (MWFeOcorr) + 159.6882
+
+
+    Fe_mol_weights = pd.DataFrame(list(zip(MWil,MWhyp, MWol, MWmt)),
+                                  columns = ['ilmenite', 'hypersthene', 'olivine','magnetite'])
+
+
+    # Add MnO to FeO
+    data['FeO'] = data['FeO'] + data['MnO']
+    data = data.drop(['MnO'], axis = 1)
+
+    # Calculate molecular weights for some more difficult minerals
+    MWAp = 3*mol_weights['CaO']+(1/3)*(mol_weights['CaO']-15.9994)+154.6101241 # apatite
+    MWNaCl = ((mol_weights['Na2O']-15.9994)/2) + 35.4527 # Halite - adding MW Cl to Na
+    MWCaF2 = (mol_weights['CaO']-15.9994) + (2*18.9984032) # Fluorite - add fluorine to Ca MWR
+    MWFeS2 = (mol_weights['FeO']-15.9994) + (2*32.066) # Pyrite - add sulphur to Fe
+
+    # Normative minerals
+    #Zircon??
     # Apatite
-
-    # Flourite
-
-    # halite
-
-    # thenardite
-
-    # pyrite
-
-    # cancrinite or calcite
-
-    # chromite
-
-    # ilmenite
-
-    # orthoclase / potassium metasilicate
-
-    # albite
-
-    # acmite / sodium metasilicate
-
-    # anorthite / corundum
-
-    # titanite / rutile
-
-    # magnetite / hematite
-
-    # diopside / wollastonite / hypersthene
-
-    # quartz / undersadurated minerals
-
-    # silica deficiency?
-
-    # olivine / hypersthene
-
-    # titanite / perobskite
-
-    # nephenline / albite
-
-    # leucite / orthoclase
-
-    # dicalcium silicate / wollastonite
-
-    # dicalcium silicate / olivine adjustment
-
-    # kaliophilite / leucite
-
-    # allocate provisional mineral proportions to definite ones
-
-    # normalise
-
-    # calculate Salic, Femic, C.I., D.I
-
-    pass
+    data['apatite'] = np.where(data['CaO']>=((3+(1/3))*data['P2O5']), data['P2O5'], data['CaO']/(3+(1/3)))
```

Comment:
For this section I'd suggest starting with an empty `norm` dataframe which you then 'fill', at the same time as extracting components from the `res` dataframe, rather than adding mineral columns to the original. Adding a bit of comment/structure will help it be a bit more readable. And you can get rid of some otherwise slightly confusing `np.where` statements:
```python
  ############################################################################
  # Calculate normative components
  ############################################################################
  # Normative minerals
  # Zircon??
  # Apatite
  norm = pd.DataFrame(index=res.index)  # dataframe to store mineralogy
  ap_fltr = res["CaO"] >= ((10.0 / 3) * res["P2O5"])  # where phosphate is limiting
  norm["apatite"] = np.where(ap_fltr, res["P2O5"], res["CaO"] / (10.0 / 3))
  res["CaO"] = res["CaO"] - (10.0 / 3) * norm["apatite"]
  res["P2O5"] = res["P2O5"] - norm["apatite"]
  # Fluorite
  # Halite
  # Thenardite
  then_fltr = res["Na2O"] >= res["SO3"]  # where sulfate is limiting
  norm["thenardite"] = np.where(then_fltr, res["SO3"], res["Na2O"])
  res["SO3"] = res["SO3"] - norm["thenardite"]
  res["Na2O"] = res["Na2O"] - norm["thenardite"]
  ...
  ```
You can also remove the `xx_fltr` lines and reduce it further if you wish, but I'd suggest leaving a few comments for readability. Note that this example is a simple case; where the relationship between components and phases isn't 1:1, you'll need to multiply by some factor like `norm["<mineral>"].values[:, None] * np.array([1, 3])[None, :]` for the second line here, so splitting it out on multiple lines still generally makes sense:
  ```
norm["thenardite"] = np.where(res["Na2O"] >= res["SO3"] , res["SO3"], res["Na2O"]) # where sulfate is limiting
res[["SO3", "Na2O"]] -= norm["thenardite"]
  ```